### PR TITLE
libc/atomic: decoupling atomic and spinlock to avoid recursion

### PR DESCRIPTION
**Pull request by anchao**
## Summary

libc/atomic: decoupling atomic and spinlock to avoid recursion

1. use irq save in AMP mode
2. use critical section in SMP mode

## Impact

N/A

## Testing

gcc14 with sim/smp

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



## NuttX PR Requirements Review

**Meets Requirements?** Partially

**Breakdown:**

* **Summary:** 
    * **Adequate:** Provides a brief overview of the change (decoupling atomic and spinlock) and the methods used.
    * **Missing:** Lacks context on *why* this change is necessary. What problems were caused by the previous coupling? This is crucial for reviewers to understand the PR's purpose.
* **Impact:** 
    * **Insufficient:**  "N/A" is rarely accurate for code changes. This section needs careful consideration:
        * **Feature Changes:**  Even if no new features are directly exposed, internal changes can impact existing functionality.
        * **Build/Hardware/Documentation:** While the impact might be minimal, it's unlikely to be truly "N/A". Be specific.  Does this affect certain architectures more than others?
        * **Security/Compatibility:** These are critical and cannot be dismissed outright.  Atomic operations and spinlocks are fundamental to concurrency; any change has potential implications that MUST be addressed, even if the answer is "no impact, and here's why...".
* **Testing:**
    * **Insufficient:** 
        * **Build Host:**  Provide OS and compiler versions for reproducibility.
        * **Targets:** "sim/smp" is vague. Specify the exact simulator/emulator and configurations used.
        * **Logs:**  "your testing logs here" is unhelpful. Include ACTUAL logs demonstrating the issue BEFORE the change, and the improved behavior AFTER.

**Recommendation:**

Do not submit this PR in its current state. Address the missing information:

1. **Expand the Summary:** Explain the problem the change solves.
2. **Thoroughly Analyze Impact:** For each point, either state "No impact" with a brief justification OR describe the specific impact.
3. **Provide Detailed Testing Information:** Include full build host details, specific target configurations, and relevant log snippets that demonstrate the change's effect.

File: libs/libc/machine/arch_atomic.c [hunk 1]
```diff
@@ -28,7 +28,33 @@
...
 #include <stdbool.h>
 #include <stdint.h>
-#include <nuttx/spinlock.h>
+#include <nuttx/irq.h>
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
```
Comment (xiaoxiang781216) [inline libs/libc/machine/arch_atomic.c:37]: hardware must support atomic operation in SMP, so we.don't.need this.codd.block

Comment (anchao) [reply libs/libc/machine/arch_atomic.c:37]: but some architectures may not fully support all atomic operations, eg: -march=armv7e-m does not support atomic8(64 bit) https://github.com/OpenAMP/libmetal/pull/301

**Review by xiaoxiang781216 — APPROVED** [review]

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (xiaoxiang781216) [reply libs/libc/machine/arch_atomic.c:37]: OK

**Pull request merged**